mlflow/types/type_hints.py

```diff
@@ -0,0 +1,309 @@
+        )
+    raise MlflowException.invalid_parameter_value(
+        f"Unsupported type hint `{type_hint}`, supported types are: "
+        f"{list(TYPE_HINTS_TO_DATATYPE_MAPPING.keys())}, pydantic BaseModel subclasses, "
+        "lists and dictionaries of primitive types, or typing.Any."
+    )
+
+
+def _infer_fields_from_pydantic_model(
+    model: pydantic.BaseModel, field_type: Literal["Property", "ColSpec"]
+) -> list[Union[Property, ColSpec]]:
+    """
+    Infer the fields from a pydantic model.
+    If field_type is "Property", the model is seen as an Object, and output fields
+    are inferred as Properties.
+    If field_type is "ColSpec", the model is seen as a schema, and output fields
+    are inferred as ColSpecs.
+    """
+    if _is_pydantic_type_hint(model):
+        fields = model_fields(model)
+    else:
+        raise TypeError(f"model must be a Pydantic model class, but got {type(model)}")
+
+    if field_type not in ["Property", "ColSpec"]:
+        raise MlflowException.invalid_parameter_value(
+            f"field_type must be 'Property' or 'ColSpec', but got {field_type}"
+        )
+
+    output_fields = []
+    invalid_fields = []
+    for field_name, field_info in fields.items():
+        annotation = field_info.annotation
+        # this shouldn't happen since pydantic has checks for missing annotations
+        # but added here to avoid potential edge cases
+        if annotation is None:
+            invalid_fields.append(field_name)
+            continue
+        colspec_type = _infer_colspec_type_from_type_hint(annotation)
+        if colspec_type.required is False and field_required(field_info):
+            raise MlflowException.invalid_parameter_value(
+                f"Optional field `{field_name}` in Pydantic model `{model.__name__}` "
+                "doesn't have a default value. Please set default value to None for this field."
+            )
+        if field_type == "Property":
+            output_fields.append(
+                Property(
+                    name=field_name,
+                    dtype=colspec_type.dtype,
+                    required=colspec_type.required,
+                )
+            )
+        elif field_type == "ColSpec":
+            output_fields.append(
+                ColSpec(
+                    type=colspec_type.dtype,
+                    name=field_name,
+                    required=colspec_type.required,
+                )
+            )
+    if invalid_fields:
+        raise MlflowException.invalid_parameter_value(
+            "The following fields in the Pydantic model do not have type annotations: "
+            f"{invalid_fields}. Please add type annotations to these fields."
+        )
+
+    return output_fields
+
+
+def _is_pydantic_type_hint(type_hint: type[Any]) -> bool:
+    try:
+        return inspect.isclass(type_hint) and issubclass(type_hint, pydantic.BaseModel)
```

Comment:
```suggestion
        return issubclass(type_hint, pydantic.BaseModel)
```
This is equivalent, right?